src/core/stringify.ts

```diff
@@ -0,0 +1,58 @@
+import { htmlEncode } from '../shared/util';
+import type { Hooks } from './hooks';
+import type { Token } from './token';
+import type { TokenStream } from './token';
+
+/**
+ * Converts the given token or token stream to an HTML representation.
+ *
+ * The following hooks will be run:
+ * 1. `wrap`: On each {@link Token}.
+ *
+ * @param o The token or token stream to be converted.
+ * @param language The name of current language.
+ * @returns The HTML representation of the token or token stream.
+ */
+function stringify (o: string | Token | TokenStream, language: string, hooks: Hooks): string {
+	if (typeof o === 'string') {
+		return htmlEncode(o);
+	}
+	if (Array.isArray(o)) {
+		let s = '';
+		o.forEach(e => {
+			s += stringify(e, language, hooks);
+		});
+		return s;
+	}
+
+	const env: Record<string, any> & { classes: string[]; attributes: Record<string, string> } = {
+		type: o.type,
```

Comment:
`o` is a string. Does it have a `type` property?

Reply:
It's not just a string. If it's a string, we've already returned in line 18.

Reply:
Oh, right. Here, it's a `TokenStream` or a `Token`. My bad.